### ISSUE TITLE
Figure out how to handle version changes

### ISSUE THREAD

**Issue by tvansteenburgh** [juju/python-libjuju#49]
See https://lists.ubuntu.com/archives/juju/2017-January/008479.html for context.

Loading the correct facade version dynamically is the way to go. The schema json file will need to be appended to instead of overwritten so that old facade versions are preserved. API classes will need to be namespaced so that two versions of the same facade (with the same name) don't collide.

That's the easy part. The tricky part is the coupling between the OO and api layers. For example, Application.add_units() calls ApplicationFacade.AddUnits() with parameters. What happens when the next version of that facade changes the parameters? The OO layer needs some way to deal with that...

**Comment by petevg** [comment]
I'm tempted to do some munging at import time. Set an environment variable (or some other flag) before importing juju, and then have client.py import _client<version>.py, depending on the flag.

Another option would be to have a set of wrappers (one per version) sitting in front of the object layer. Depending on which wrapper you imported your objects from, we'd do something different with the _client import. That would break if people tried to dig deeper into the library for their imports, though.

**Comment by petevg** [comment]
@tvansteenburgh @johnsca @abentley I have a proposal for a solution, with some psuedo code to illustrate how it might work.

It lives in a pastebin: http://paste.ubuntu.com/24320954/

What do you think. Is it a workable solution?

**Comment by johnsca** [comment]
@petevg I'm not a big fan of having the implementation be dynamically loaded at run-time.  It makes it hard to know what params to pass to the facades, whether by introspection, reading the source, or reading generated docs.  Also, if / when https://bugs.launchpad.net/juju/+bug/1679438 happens, we should be using that schema definition directly and shouldn't even need schemagen or schema JSON files in the libjuju repo anymore.

Instead, I'd prefer to split `_client.py` into separate packages for each facade with a file for each version of the facade but still have them pre-generated.  We could even have an optimization that only looks for new versions of facades that don't have implementations and don't bother regenerating existing versions.  I think this gives us a natural place to put generated docstrings, makes the params very explicit, and should give us nice commits that make it very clear what's new.  It would also make it easy to diff two versions of a facade to see what changed.  I think the import logic would work pretty much how you described for loading the versioned schema in the pastebin.

Does that seem like a reasonable counter-proposal?

**Comment by johnsca** [comment]
Something that neither proposal addresses is how to handle the different facade versions in the object layer.  I think "facade" is an inaccurate term for the classes that are being generated, and they're actually versioned API interfaces.  In a sense, the object layer is our facade, though it also contains application logic in a lot of places.  It might make sense to split that and have the object layer contain all of the logic and a separate proper facade layer that does nothing but translate model objects to the proper versioned API calls.

**Comment by petevg** [comment]
@johnsca That seems reasonable, and was, in fact, the implementation I settled on over breakfast this morning. :-)

I switched to generating the Python classes dynamically because my pseudo code started to get messy.

To import the right client, you do something like:

```
__import__("_client{}.py".format(version)) as client
```

And then you `getattr(client, self.__name__)`, and have some annoying variable names to juggle.

That mess is all contained in the .connection routine, however, and it's a lot nicer to read Python than it is to read .json. I think that I am +1 to writing out _client<version>.py files, and I will do my best not to commit too many sins when implementing the patcher ...

**Comment by petevg** [comment]
@johnsca Actually, the other messy bit is that if you have stuff written in a json doc, it's easy to see which routines you need to monkey patch. If you're trying to slam a Python class on another Python class, it's a bit messier. 

Though I could get probably get around that with doing something tricky with inheritance, rather than using setattr. That's actually a strong argument in favor of the factory pattern ...

**Comment by abentley** [comment]
I am not a fan of the ```connect``` method.  You always want your object to be connected, so you should be able to assume it's connected.

Instead of 
```python
    action = ActionFacade()
    action.connect(connection)
```

I would like to do
```python
    action = ActionFacade.from_connection(connection)
```

This allows the ```from_connection``` factory method to determine what specific facade is appropriate from the connection.  Depending on the implementation, it can either return a different ActionFacade subclass or initialize ActionFacade differently.  And it means we don't have to worry about uninitialized ActionFacades.

**Comment by petevg** [comment]
@abentley Agreed. The argument for using .connect is that the whole library already calls connect, so you'd need to do minimal refactoring.

Changing a bunch of calls from a two liner including .connect to a one liner including .from_connection isn't too bad, though, and we've been discouraging people from using the Facade classes outside of python-libjuju, so I don't think that we'll necessarily break anybody.

I think that this, combined with @johnsca's push for pre-generating _client<version>.py is a big +1 for the factory pattern. I might even be able to avoid a single call to getattr and setattr outside of the initial generation of the _client<version>.pys :-)

**Comment by petevg** [comment]
Updated spec, based on feedback: http://paste.ubuntu.com/24323064/

**Comment by tvansteenburgh** [comment]
The controller or model tells you every facade+version that it supports in response to the Login api request.

**Comment by tvansteenburgh** [comment]
> Something that neither proposal addresses is how to handle the different facade versions in the object layer.

This is the most important part IMO. Loading the correct facade version based on what the api server tells you it supports is straightforward. The trickier part is making the object layer able to use any version of a facade internally, even when the interface of that facade changes between versions.

**Comment by abentley** [comment]
If you want to present the same python API for two different facade versions, you're looking at a lowest-common-denominator API.  If a newer Juju supports the older facade, the simplest thing is to use the older facade version for all Jujus and ignore the newer one.

If you want the object-layer API for the newer facade to be a superset of the object-layer API for the older facade, you'll probably need a different class for each facade version, using inheritance to add additional capabilities and override some implementations.

**Comment by petevg** [comment]
> The controller or model tells you every facade+version that it supports in response to the Login api request.

@tvansteenburgh That makes things easier! (From my initial read on the docs, I didn't realize that the version info it gave back was complete.) Will tweak the proposal to match ...

> The trickier part is making the object layer able to use any version of a facade internally, even when the interface of that facade changes between versions.

Right now, it would be nice to just be able to run against 2.2 and 2.0.3 (for JaaS) without breaking because a key that we don't use has been renamed. I agree that we need a formal system (or at least a battle plan) to act as an intermediary between the business logic in the object layer and the ~~facades~~versioned API interfaces, though.

I think that what I'd like to do is implement my proposal right now (tweaked to use facade versions rather than application versions), so that we unblock conjure-up and juju-matrix, and let both projects hit their deadlines this month, and then move on to implementing @johnsca's proper facades, probably via the subclassing routine that @abentley suggested (perhaps using a factory pattern similar to the way that we're fetching the right api interface.)

Does that sound like a plan to everyone?

**Comment by abentley** [comment]
It sounds good to me.

**Comment by petevg** [comment]
Updated spec, just for the record: http://paste.ubuntu.com/24327802/  Thx for the comments, all :-)
